File: app/src/main/java/com/pawels96/skyrimperkcalculator/ui/SkillTreeFragment.java
package com.pawels96.skyrimperkcalculator.ui;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.app.AlertDialog;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.TextView;

import com.pawels96.skyrimperkcalculator.R;
import com.pawels96.skyrimperkcalculator.enums.SkillEnum;
import com.pawels96.skyrimperkcalculator.models.Perk;

import static android.content.Context.LAYOUT_INFLATER_SERVICE;
import static com.pawels96.skyrimperkcalculator.Utils.getPerkDescription;
import static com.pawels96.skyrimperkcalculator.Utils.getPerkName;


public class SkillTreeFragment extends Fragment {

    private OnFragmentInteractionListener mListener;
    private GraphView graphView;

    public static final String ARG_1 = "arg1";

    private MainActivity parentActivity;
    private TextView activePerks, reqSkill;

    public SkillTreeFragment() {}

    private SkillEnum skillEnum;

    public interface OnFragmentInteractionListener {
        void onPerkChanged(SkillEnum skill, Perk perk, int state);
    }

    public static SkillTreeFragment newInstance(SkillEnum param1) {
        SkillTreeFragment fragment = new SkillTreeFragment();
        Bundle args = new Bundle();
        args.putSerializable(ARG_1, param1);
        fragment.setArguments(args);
        return fragment;
    }

    public String getPerkCount() {
        return graphView.getSkill().getPerksCount();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        skillEnum = (SkillEnum)getArguments().getSerializable(ARG_1);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View rootView = inflater.inflate(R.layout.fragment_skilltree, container, false);
        graphView = rootView.findViewById(R.id.graph);

        activePerks = rootView.findViewById(R.id.skillPerks);
        reqSkill = rootView.findViewById(R.id.skillLevel);

        graphView.setListener(listener);
        graphView.setSkill(parentActivity.getBuild().getSkill(skillEnum), parentActivity.getBuild().getPerkSystem());
        updateSkillInfo();

        return rootView;
    }

    public void refresh() {
        graphView.setSkill(parentActivity.getBuild().getSkill(skillEnum), parentActivity.getBuild().getPerkSystem());
        updateSkillInfo();
        graphView.invalidate();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        parentActivity = (MainActivity) context;
        mListener = (OnFragmentInteractionListener) context;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    private GraphView.OnNodeClickedListener listener = new GraphView.OnNodeClickedListener() {
        @Override
        public void onNodeClicked(Perk perk) {
            mListener.onPerkChanged(skillEnum, perk, perk.getState());
            updateSkillInfo();
        }

        @Override
        public void onNodePressed(Perk perk) {
            showPerkDescription(perk);
        }
    };

    private void showPerkDescription(Perk perk){

        final AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(getActivity(), R.style.CustomDialogTheme);

        LayoutInflater inflater = (LayoutInflater) getActivity().getSystemService(LAYOUT_INFLATER_SERVICE);
        View customView = inflater.inflate(R.layout.popup_perk_description, null);
        dialogBuilder.setView(customView);

        TextView perkDesc = customView.findViewById(R.id.perk_desc);
        TextView perkSkill = customView.findViewById(R.id.perk_skill);

        String skillText = getString(R.string.required_skill) + ": " + perk.getAllSkillLevels();
        perkSkill.setText(skillText);
        perkDesc.setText(getPerkDescription(getActivity(), perk.getPerk()));

        dialogBuilder.setTitle(getPerkName(getActivity(), perk.getPerk()));
        dialogBuilder.setCancelable(true);

        final AlertDialog dialog = dialogBuilder.create();

        Window window = dialog.getWindow();
        if (window != null)
        window.setGravity(Gravity.BOTTOM);

        dialog.show();
    }

    private void updateSkillInfo() {

        String perksText = getString(R.string.active_perks) + ": " + getPerkCount();
        String skillText = getString(R.string.required_skill) + ": "  + graphView.getSkill().getRequiredSkillLevel();

        activePerks.setText(perksText);
        reqSkill.setText(skillText);
    }

}

File: app/src/main/java/com/pawels96/skyrimperkcalculator/presentation/SkillTreeFragment.kt
package com.pawels96.skyrimperkcalculator.presentation

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import com.pawels96.skyrimperkcalculator.Injector
import com.pawels96.skyrimperkcalculator.R
import com.pawels96.skyrimperkcalculator.databinding.FragmentSkilltreeBinding
import com.pawels96.skyrimperkcalculator.domain.*
import com.pawels96.skyrimperkcalculator.presentation.dialogs.PerkInfoDialog
import com.pawels96.skyrimperkcalculator.presentation.viewmodels.BuildsViewModel
import com.pawels96.skyrimperkcalculator.presentation.views.GraphView.OnNodeClickedListener

class SkillTreeFragment : Fragment(R.layout.fragment_skilltree) {

    private val binding by viewBinding(FragmentSkilltreeBinding::bind)

    private lateinit var displayedSkill: ISkill

    private val model: BuildsViewModel by lazy {
        ViewModelProvider(
            requireActivity(),
            Injector.provideVmFactory()
        )[BuildsViewModel::class.java]
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        displayedSkill = requireArguments().getSerializable(ARG_SKILL, ISkill::class.java) as ISkill
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.graph.listener = listener
        model.currentBuild.observe(viewLifecycleOwner) { displayBuild(it) }
    }

    private fun displayBuild(build: Build) {
        val skill = build.getSkill(displayedSkill)
        val perksText =
            getString(R.string.active_perks) + ": ${skill.selectedPerksCount}/${skill.maxPerks}"
        val requirementLabelRes =
            if (skill is SpecialSkill) R.string.required_kills else R.string.required_skill
        val skillText = getString(requirementLabelRes) + ": " + skill.requiredSkillLevel
        binding.graph.display(skill)
        binding.skillPerks.text = perksText
        binding.skillLevel.text = skillText
    }

    fun cancelHold() {
        binding.graph.cancelHold()
    }

    private val listener: OnNodeClickedListener = object : OnNodeClickedListener {
        override fun onNodeClicked(perk: Perk) {
            model.changePerkState(displayedSkill, perk.perk)
        }

        override fun onNodeHolding(perk: Perk) {
            showPerkDescription(perk)
        }
    }

    private fun showPerkDescription(perk: Perk) {
        PerkInfoDialog.create(displayedSkill, perk).show(childFragmentManager)
    }

    companion object {

        private const val ARG_SKILL = "arg_skill"

        fun create(skill: ISkill): SkillTreeFragment {
            return SkillTreeFragment().apply {
                arguments = Bundle().apply {
                    putSerializable(ARG_SKILL, skill)
                }
            }
        }
    }
}

